[ROCm] StreamExecutor logic for ROCm platform (PR 20709 / 22669 continued)

deven-amd opened the pull request:
/cc @timshen91 

hi Tim

This is a continuation of the PR #22669

This PR contains the changes you requested in your comments in PR #22669. (specifically the changes to have common header files for CUDA/ROCm in the `tensorflow/stream_executor/gpu/` directory)

The changes are broken down into multiple commits to make ti easier to code review (as per your request). Once the review process in done, and we are ready to merge, we can squash commits into fewer / one.

I have `git mv`'d the header files from the stream_executor/cuda to the stream_executor/gpu dir, to preserve their change history

For the time being I have left in the CUDA specific *comments* in the header files within `stream_executor/gpu` dir. Let me know if it is okay to remove them.

Thanks

deven

timshen91 commented on tensorflow/stream_executor/rocm/rocm_activation.cc:
```diff
@@ -0,0 +1,48 @@
+namespace stream_executor {
+namespace gpu {
+
+GpuContext* ExtractGpuContext(GpuExecutor* rocm_exec);
```
This file and cuda_activation.cc are identical except for variable names. Are they going to diverge in the future? Maybe just merge them?

deven-amd replied:
will merge `rocm/rocm_activation.cc` and `cuda/cuda_activation.cc` into a common `gpu/gpu_activation.cc` in the next commit.

I did not realize that you were okay with merging the .cc files as well (as opposed to just the header files)

timshen91 commented on tensorflow/stream_executor/rocm/rocm_event.cc:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  switch (status.ValueOrDie()) {
+    case hipSuccess:
```
ditto: this file is largely the same as cuda_event, except for PollForStatus. Maybe merge the rest of the functions?

deven-amd replied:
will create a new `gpu/gpu_event.cc` which will contain all the routines except for `PollForStatus`. That routine will continue to stay in `rocm/rocm_event.cc` / `cuda/cuda_event.cc`. 

The `PollForStatus` routine will be only content in the `rocm/rocm_event.cc` / `cuda/cuda_event.cc` files. Going forward the idea will be to keep routines that have common implementation in the `gpu/gpu_*.cc` files, and ROCm/CUDA specific implementation in the `rocm/rocm_*.cc` / `cuda/cuda_*.cc` files.

timshen91 commented on tensorflow/stream_executor/rocm/rocm_stream.cc:
```diff
@@ -0,0 +1,61 @@
+  return GpuDriver::IsStreamIdle(parent_->gpu_context(), gpu_stream_);
+}
+
+GpuStream *AsGpuStream(Stream *stream) {
```
This file is trivially similar to cuda_stream as well.

deven-amd replied:
will merge `rocm/rocm_stream.cc` and `cuda/cuda_stream.cc` into a common `gpu/gpu_stream.cc` in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_timer.cc:
```diff
@@ -0,0 +1,86 @@
+namespace stream_executor {
+namespace gpu {
+
+bool GpuTimer::Init() {
```
ditto: This file is a bit different from cuda_timer, but I didn't see any logical difference.

deven-amd replied:
will merge `rocm/rocm_timer.cc` and `cuda/cuda_timer.cc` into a common `gpu/gpu_timer.cc` in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_diagnostics.cc:
```diff
@@ -0,0 +1,238 @@
+namespace gpu {
+
+string DriverVersionToString(DriverVersion version) {
+  return port::Printf("%d.%d.%d", std::get<0>(version), std::get<1>(version), std::get<2>(version));
```
For all `port::Printf()` calls, can you use `absl::StrFormat` instead? One immediate benefit is to get rid of all `c_str()`s

deven-amd replied:
will be fixed in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_diagnostics.cc:
```diff
@@ -0,0 +1,238 @@
+port::StatusOr<DriverVersion> StringToDriverVersion(const string &value) {
+  std::vector<string> pieces = port::Split(value, '.');
+  if (pieces.size() != 2 && pieces.size() != 3) {
+    return port::Status{
```
`return InvalidArgument(...);`

so that it's more consistent with other tensorflow files.

Please check all places.

deven-amd replied:
is it okay to reach into `tensorflow::errors::InvalidArgument` from within the stream_executor code?
 
Access to `Status` from within stream_executor code seems to be routed through `stream_exectuor::port` namespace (see `stream_executor/lib/status.h`). 

Can we leave this part as is?

timshen91 commented on tensorflow/stream_executor/rocm/rocm_driver.cc:
```diff
@@ -0,0 +1,1356 @@
+
+///////////////
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wswitch"
```
Can you document which GCC warnings it's trying to ignore?

deven-amd replied:
the pragmas are no longer needed, they will be gone in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_driver.cc:
`if (value32 != value) {
  LOG(ERROR) << "...";
  return false;
}`?

deven-amd replied:
will be fixed in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_driver.cc:
ditto.

deven-amd replied:
will be fixed in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_driver.cc:
TODO(ROCM): properly implement this feature in HIP

Same for other places.

deven-amd replied:
will be fixed in the next commit

timshen91 commented on tensorflow/stream_executor/rocm/rocm_gpu_executor.cc:
```diff
@@ -0,0 +1,978 @@
+  auto& module = module_it->second.first;
+  auto& refcount = module_it->second.second;
+  VLOG(3) << "Found HSACO module " << module << " with refcount " << refcount;
+  if (--refcount == 0) {
```
Just curious, do you need a lock for the reference counting?

whchung replied:
`UnloadModule` is protected by a lock and the `UnloadGpuBinary` is called so we don't really a lock here.

timshen91 commented on tensorflow/stream_executor/rocm/rocm_rng.cc:
```diff
@@ -0,0 +1,310 @@
+}  // namespace wrap
+
+template <typename T>
+string TypeString();
```
TypeString should be shared with cuda.

deven-amd replied:
will move `TypeString` templates to the `gpu/gpu_rng.h` header file

timshen91 commented on tensorflow/stream_executor/rocm/rocm_rng.h:
```diff
@@ -0,0 +1,100 @@
+
+namespace stream_executor {
+
+class Stream;
```
Can we merge this file with cuda_rng.h?

deven-amd replied:
will merge `rocm/rocm_rng.h` and `cuda/cuda_rng.h` into a common `gpu/gpu_rng.h` in the next commit

timshen91 commented:
Thanks Deven, this PR looks much better!

Overall the commits look good, I have some comments in the second last one, regarding merging more files.

One last thing: can you `git clang-format` all files in the end?

deven-amd commented on tensorflow/stream_executor/cuda/cuda_driver.cc:
```diff
@@ -138,11 +138,22 @@ string ToString(CUresult result) {
 
+// Returns the current context set in CUDA. This is done by calling the cuda
+// driver (e.g., this value is not our cached view of the current context).
+static CUcontext CurrentContextOrDie() {
```
will be fixed in the next commit

deven-amd commented on tensorflow/stream_executor/gpu/gpu_executor.h:
```diff
@@ -71,11 +73,13 @@ class GpuExecutor : public internal::StreamExecutorInterface {
               const BlockDim &block_dims, const KernelBase &k,
               const KernelArgsArrayBase &args) override;
 
+  // (supported on CUDA only)
```
I thought about this, but this change is non-trivial and might not be worth it.

Code outside of  stream_executor does not call routines that are CUDA/ROCm specific, but those routines are called by other routines which are called from outside. Take for example `GpuExecutor::LoadModule`. It gets called from the XLA code, and in turn will call `LoadPtx / LoadCubin / LoadHsaco`. 

If I move the `LoadHsaco` routine to a `ROCmExecutor` class, then I need to introduce a `reinterpret_cast<ROCmExecutor*>(this)->LoadHsaco` in the implementation of `GpuExecutor::LoadModule`. Potentially lot of similar casts will need to be added. 

`LoadHsaco` needs access to the `in_memory_modules_mu_` mutex, which is also used by other routines, so basically we will end up moving all the private members of `GpuExecutor` to `ROCmExecutor / CUDAExecutor` (or declaring some  as `protected` and leaving them in `GpuExecutor`....that will make code difficult to follow and maintain)

 At that point we might as well revert back to having separate classes / header files for `rocm_gpu_executor.h` / `cuda_gpu_executor.h`. But doing so will require undoing some of the header files that we have merged (since they refer to GpuExecutor)

What do you think?

timshen91 replied:
I think it's ok for now, but maybe something we want to revisit in the future when the divergence gets larger.

deven-amd commented on tensorflow/stream_executor/gpu/gpu_driver.h:
```diff
@@ -205,14 +207,22 @@ class GpuDriver {
 
   // Loads ptx_contents with the CUDA driver's PTX JIT and stores the resulting
   // handle in "module". Any error logs that are produced are logged internally.
+  // (supported on CUDA only)
```
see comment above

deven-amd commented on tensorflow/stream_executor/gpu/gpu_executor.h:
```diff
@@ -256,23 +269,41 @@ class GpuExecutor : public internal::StreamExecutorInterface {
       EXCLUSIVE_LOCKS_REQUIRED(in_memory_modules_mu_);
 
+  // Guards the on-disk-module mapping.
+  mutex disk_modules_mu_;
```
see comment above

deven-amd commented:
@timshen91 

Hi Tim,

I have pushed commits that address some of the changes you requested. For the rest, I need further input from you.

I will `git clang-format` all the changes once we have resolve all the issues.

Thanks

deven

rmlarsen commented:
@deven-amd could you please resolve the conflict?

deven-amd commented:
@timshen91 @rmlarsen 

I will push out a rebase shortly to resolve the conflict

deven-amd commented:
hmm....I did not request a review from `azaks2` and `chsigg` ... do not know what is going on.

deven-amd commented:
@timshen91 any update on this PR?

timshen91 commented:
@deven-amd Sorry, I'll get back to the PR in one day.

timshen91 commented:
Looks good. Please resolve the conflicts and clang-format the files.

deven-amd commented:
hi Tim, 

After I did the rebase, I am running into the following error when I try to do a TF build (with --config=cuda) 

```
ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:29:1: @local_config_cuda//crosstool:cc-compiler-local: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:48:1: @local_config_cuda//crosstool:cc-compiler-darwin: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:63:1: @local_config_cuda//crosstool:cc-compiler-windows: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
```

The changes in this PR do not touch the CUDA specific settings in the BAZEL build files, so I do not know what is causing this error. Only thing I can think off is that I picked up some commit with the rebase that breaks the build.

deven

timshen91 commented:
Removing @azaks2 and @chsigg unless they also want to review this (which seems not intended).

timshen91 commented:
> hi Tim,
> 
> After I did the rebase, I am running into the following error when I try to do a TF build (with --config=cuda)
> 
> ```
> ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:29:1: @local_config_cuda//crosstool:cc-compiler-local: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
> ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:48:1: @local_config_cuda//crosstool:cc-compiler-darwin: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
> ERROR: /root/.cache/bazel/_bazel_root/efb88f6336d9c4a18216fb94287b8d97/external/local_config_cuda/crosstool/BUILD:63:1: @local_config_cuda//crosstool:cc-compiler-windows: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
> ```
> The changes in this PR do not touch the CUDA specific settings in the BAZEL build files, so I do not know what is causing this error. Only thing I can think off is that I picked up some commit with the rebase that breaks the build.
> 
> I get the same error with the TF build with `--config==rocm` as well.
> Both ROCm and CUDA builds working fine before the rebasing.
> 
> deven

I was unaware of such an error. I'll take a look later.

deven-amd commented:
I get the same error, when I try to a TF build (with --config=cuda) using the tensorflow repo as well....something is broken on the tip :(

```
root@3cb559125c7d:~/google_tensorflow# git log -1
commit c0b2e3eb7c2c02b3725bdda834e7b5d2875e1cf0
Author: A. Unique TensorFlower <gardener@tensorflow.org>
Date:   Wed Dec 12 17:00:06 2018 -0800

    Adds unicode_decode and unicode_decode_with_offsets ops, which decode strings into unicode codepoints.
    
    Adds unicode_split and unicode_split_with_offset ops, which split strings into unicode characters.
    
    RELNOTES: Adds unicode_decode, unicode_decode_with_offsets, unicode_split, and unicode_split_with_offset ops.
    PiperOrigin-RevId: 225281768


root@3cb559125c7d:~/google_tensorflow# bazel build --config=opt --config=cuda //tensorflow/tools/pip_package:build_pip_package --verbose_failures
WARNING: The following configs were expanded more than once: [cuda]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
WARNING: --batch mode is deprecated. Please instead explicitly shut down your Bazel server using the command "bazel shutdown".
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: @local_config_cuda//crosstool:cc-compiler-local: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:48:1: @local_config_cuda//crosstool:cc-compiler-darwin: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:63:1: @local_config_cuda//crosstool:cc-compiler-windows: no such attribute 'toolchain_identifier' in 'cc_toolchain' rule
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:83:1: Target '@local_config_cuda//crosstool:clang/bin/crosstool_wrapper_driver_is_not_gcc' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:crosstool_wrapper_driver_is_not_gcc'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:crosstool_wrapper_driver_is_not_gcc' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:crosstool_wrapper_driver_is_not_gcc' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/crosstool/BUILD:29:1: Target '@local_config_cuda//crosstool:empty' contains an error and its package is in error and referenced by '@local_config_cuda//crosstool:cc-compiler-local'
ERROR: /root/.cache/bazel/_bazel_root/dc5f19088b30784325d5a276b57d6400/external/local_config_cuda/cuda/BUILD:40:1: every rule of type cc_library implicitly depends upon the target '@local_config_cuda//crosstool:cc-compiler-local', but this target could not be found because of: Target '@local_config_cuda//crosstool:cc-compiler-local' contains an error and its package is in error
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Analysis failed
INFO: Elapsed time: 3.024s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (61 packages loaded)
    currently loading: tensorflow/python ... (5 packages)
```

deven-amd commented:
figure out the cause of my build errors...it was the bazel version.

I was using version `0.15.0` based upon the version used in the tensorflow repo here 
(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/install/install_bazel.sh#L18)

Moving up the bazel version to `0.19.2` seems to fix the build errors shown above. Still do not know why this update to bazel is required, but I am happy that the build is working again.

There seems to have been some garbage chars introduced in one file by my previous commit (to clang-format the changes), which results in a build error. I will be pushing out a commit shortly to fix this.

deven-amd commented:
```
@deven-amd deven-amd dismissed timshen91’s stale review via 70c2bd8 4 hours from now
```

don't know whats going on with github...that is the second time in this PR it has attributed to me something I did not do! (first being adding 2 reviewers)

deven-amd commented:
@timshen91 any update on the status of this PR? 

thanks.

deven-amd commented:
@timshen91  Thank you for the quick update. Please let me know if you need anything done on my end to help you out.

deven

deven-amd commented:
@timshen91 

I have rebased the PR today to resolve merge conflicts that have sneaked in over the last couple of weeks. just FYI.

timshen91 commented:
@deven-amd , much appreciated on the rebases.

I committed https://github.com/tensorflow/tensorflow/commit/530d1b267f124fdb7dd6ce65b94423b8801152cd which makes our internal integration of github PRs much easier from now on. Can you rebase onto it?

whchung commented:
@timshen91 , @deven-amd is on vacation and I'll rebase this PR. IIRC it's not possible to have 1 PR with multiple authors, so I'll likely have to squash all of his earlier commits into 1 and force push with my github handle.

whchung commented on tensorflow/stream_executor/BUILD:
```diff
@@ -124,3 +132,37 @@ cc_library(
+        "@local_config_rocm//rocm:rocm_headers",
+    ] + if_rocm_is_configured([
+        "//tensorflow/core:rocm",
+        "@local_config_rocm//rocm:hiprand",
```
@yifeif If I understand your questions correctly, you want to have these dependencies removed from "rocm_platform" cc_library target, correct?

whchung replied:
@yifeif I've already changed the bazel script so these dependent libraries would be submitted in future PRs. Also they would become plugins and be loaded dynamically.

yifeif replied:
Sounds good to me. Thanks @whchung!

whchung commented:
@timshen91 Just to give you a heads up that I'm refactoring StreamExecutor-related bazel files to accommodate your changes in commit 530d1b2 , as well as @deven-amd 's proposed changes in this PR. Basically we'll need to introduce new bazel files for targets under `gpu` directory, which is shared on both GPU platforms, and change those in `cuda` and `rocm` directory accordingly. I'll push a couple of more commits once they are ready.

gunan commented:
To document CLA check:
CLA bot was failing due to PR having commits from multiple authors.

whchung commented:
> To document CLA check:
> CLA bot was failing due to PR having commits from multiple authors.

@gunan Thanks. Let me resolve build / test issues on CUDA path first. I'll then revise the commit history so there is one single author for this PR to tame CLA checker.

timshen91 commented:
@whchung one more thing, can you break `gpu_stream_executor_headers` into multiple, fine-grained targets, just like the rest of them?

whchung commented:
@timshen91 I've revised the bazel build targets per your suggestion. Could you take another look? Thanks.

whchung commented:
It looks like StreamExecutor for CUDA has changed a bit yesterday. Let me do a rebase.

whchung commented:
It has been discovered there was a buggy commit in upstream TensorFlow wrt location of `nvcc`. A separate PR ( #24942) has been filed to address that. Logic in this PR also has been revised to reflect such change before #24942 can be merged.

timshen91 commented:
I'm working on the integrating but it takes time, as I have to split the single commit to smaller ones. Specifically, we still have internal uses of the cuda/ headers and stream_executor::cuda namespace.

timshen91 commented:
Ok, the integration appears to be very challenging especially with all the new changes coming in. I learned some lessons.

It will be much easier for you to split this commit to multiple PR. The first PR:
* only contains changes in `stream_executor/...`.
* does not remove any `stream_executor/cuda/*.h`, so that things outside of stream_executor don't break. All the types and functions in the namespace cuda now alias to namespace gpu counterparts. For example, `namespace cuda { using CUDADriver = gpu::GpuDriver; }`.
* all `stream_executor/gpu/BUILD` targets should be only visible to `//third_party/tensorflow/stream_executor:__subpackages__`.
* target `stream_executor/gpu:X` should be only used by `stream_executor/cuda:cuda_X` or `stream_executor/rocm:rocm_X`, not `cuda_Y`. For example, `cuda:cuda_platform` should depend on `cuda:cuda_driver`, not `gpu:gpu_driver`.

The general idea is to make redundancy during transition (e.g. keep namespace cuda, but alias to gpu), and change as little as possible. I tried to ingrate the whole thing in order to "get this over with" but that simply doesn't work. It's much better to commit them incrementally.

whchung commented:
@timshen91 Let me see how to break this PR into smaller chunks per your suggestion, after rebase.

whchung commented:
@timshen91 it seems there are quite a few in-flight changes in StreamExecutor recently. To better cope with them could you help briefly summarize them? I can see that you added some `if_cuda_is_configured()` checks in CUDA-specific `bazel` targets, and also there seems to be an effort to make CUDA runtime be dynamically loaded. Is there any others we should take into consideration?

whchung committed:
renaming namespace stream_executor::cuda to stream_executor::gpu

moving class CudaContext definition from cuda_driver.h to cuda_driver.cc

making routine CurrentContextOrDie static to the cuda_driver.cc file

renaming CudaContext as GpuContext

introducing the gpu_type.h header file

replacing references to CU* types in the cuda_driver.h to corresponding Gpu* types from gpu_types.h.  The Gpu* types in gpu_types.h will resolve CU* types for the CUDA platorform and hip* types for the ROCm platform

renaming CUDADriver as GpuDriver

moving cuda/cuda_driver.h to gpu/gpu_driver.h

renaming CUDAExecutor as GpuExecutor

replacing references to CU* types in the cuda_gpu_executor.h to corresponding Gpu* types from gpu_types.h.  The Gpu* types in gpu_types.h will resolve CU* types for the CUDA platorform and hip* types for the ROCm platform

moving cuda/cuda_gpu_executor.h to gpu/gpu_executor.h

renaming CUDAEvent to GpuEvent and replacing references to CU* with corresponding Gpu* in cuda_event.h

moving cuda/cuda_event.h to gpu/gpu_event.h

renaming CUDAKernel to GpuKernel and replacing references to CU* with corresponding Gpu* in cuda_kernel.h

moving cuda/cuda_kernel.h to gpu/gpu_kernel.h

renaming CUDAStream to GpuStream and replacing references to CU* with corresponding Gpu* in cuda_stream.h

moving cuda/cuda_stream.h to gpu/gpu_stream.h

renaming CUDATimer to GpuTimer and replacing references to CU* with corresponding Gpu* in cuda_timer.h

moving cuda/cuda_timer.h to gpu/gpu_timer.h

moving cuda/cuda_activation.h to gpu/gpu_activation.h and cuda/cuda_diagnostics.h to gpu/gpu_diagnostics.h

renaming CUDAMemory*/CUDAComplex* to GpuMemory*/GpuComplex* and replacing references to CU* with corresponding Gpu* in cuda_helpers.h

moving cuda/cuda_helpers.h to gpu/gpu_helpers.h

fix for CUDA build errors introduced via rebasing

adding bazel build and related updates for stream_executor/rocm

ROCm platform specific changes to header files in stream_executor/gpu dir. Also fixing some CUDA specific references that were mistakenly left in those header files

ROCm platform stream_executor implementation (initial commit)

fix for a CUDA build error introduced by the previous commit

Moving .cc files from stream_executor/cuda to stream_executor/gpu

1. cuda/cuda_activation.cc to gpu/gpu_activation.cc (rocm/rocm_activation.cc is deleted)
2. cuda/cuda_stream.cc     to gpu/gpu_stream.cc     (rocm/rocm_stream.cc is deleted)
3. cuda/cuda_timer.cc      to gpu/gpu_timer.cc      (rocm/rocm_timer.cc is deleted)

A new file gpu/gpu_event.cc is being added, which contains the common contents from cuda/cuda_event.cc and rocm/rocm_event.cc. The cuda/rocm specific files will only contain routines whose implementations are platform specific

merging cuda/cuda_rng.h and rocm/rocm_rng/h into gpu/gpu_rng.h

minor updates requested in the PR 24156 feedback

clang-formatting the changes

removing some garbage chars at the beginning of the file, that were inadvertently introduced by the previous commit

fixing compile errors in the --config=cuda build, that were introduced by the rebasing

fixing compile errors in the --config=rocm build, that were introduced by the rebasing

Address clang-format checks

Refactor StreamExecutor bazel scripts

- Change CUDA-specific bazel targets to become GPU platform neutral ones
- Add ROCm-specific bazel targets
- Other changes in XLA bazel scripts
- Other changes in TensorFlow operator bazel scripts

whchung commented:
Closing this PR due to repeated running into issues with CLA. Will submit a new one.
